File: src/main/java/br/com/correios/enderecador/dao/CEPFactory.java
package br.com.correios.enderecador.dao;

public abstract class CEPFactory {
    public static final int DNEC = 1;
    public static final int GPBE = 2;

    public static CEPFactory factory(int type) {
        switch (type) {
            case DNEC:
                return DnecCEPFactory.getIntance();
            case GPBE:
                return GpbeCEPFactory.getIntance();
        }
        throw new IllegalArgumentException();
    }

    public abstract InterfaceEnderecoDao getEndereco();
}

File: src/main/kotlin/br/com/correios/enderecador/dao/CEPFactory.kt
package br.com.correios.enderecador.dao

import java.lang.IllegalArgumentException

abstract class CEPFactory {
    abstract val endereco: InterfaceEnderecoDao?

    companion object {
        private const val DNEC = 1
        private const val GPBE = 2
        @JvmStatic
        fun factory(type: Int): CEPFactory? {
            when (type) {
                DNEC -> return DnecCEPFactory.intance
                GPBE -> return GpbeCEPFactory.intance
            }
            throw IllegalArgumentException()
        }
    }
}
